// @microsoft/1ds-post-js/src/Constants.ts
export const Method = "POST";
export const DisabledPropertyName: string = "Microsoft_ApplicationInsights_BypassAjaxInstrumentation";

export const strDropped = "drop";
export const strSending = "send";
export const strRequeue = "requeue";
export const strResponseFail = "rspFail";
export const strOther = "oth";

export const defaultCacheControl = "no-cache, no-store";
export const defaultContentType = "application/x-json-stream";
export const strCacheControl = "cache-control";
export const strContentTypeHeader = "content-type";
export const strKillTokensHeader = "kill-tokens";
export const strKillDurationHeader = "kill-duration";
export const strKillDurationSecondsHeader = "kill-duration-seconds";
export const strTimeDeltaHeader = "time-delta-millis";
export const strClientVersion = "client-version";
export const strClientId = "client-id";
export const strTimeDeltaToApply = "time-delta-to-apply-millis";
export const strUploadTime = "upload-time";
export const strApiKey = "apikey";
export const strMsaDeviceTicket = "AuthMsaDeviceTicket";
export const strAuthXToken = "AuthXToken";
export const strSdkVersion = "sdk-version";
export const strNoResponseBody = "NoResponseBody";

export const strMsfpc = "msfpc";
